supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala

```diff
@@ -59,6 +59,85 @@ case object Cancelled extends TesRunStatus {
+       |
+       |# Remove the leading and trailing quotes
+       |BEARER_TOKEN="$${BEARER_TOKEN#\\"}"
+       |BEARER_TOKEN="$${BEARER_TOKEN%\\"}"
+       |
+       |# Use the precomputed endpoint from cromwell + WSM to acquire a sas token
+       |sas_response_json=$$(curl -s \\
+       |                    --retry 3 \\
+       |                    --retry-delay 2 \\
+       |                    -X POST "$getSasWsmEndpoint" \\
+       |                    -H "Content-Type: application/json" \\
+       |                    -H "accept: */*" \\
+       |                    -H "Authorization: Bearer $${BEARER_TOKEN}")
+       |
+       |# Store token as environment variable
+       |export $environmentVariableName=$$(echo "$${sas_response_json}" | jq -r '.token')
+       |
+       |# Echo the first characters for logging/debugging purposes. "null" indicates something went wrong.
+       |echo Acquired sas token: "$${$environmentVariableName:0:4}****"
+       |### END ACQUIRE LOCAL SAS TOKEN ###
+       |""".stripMargin
+  }
+
+  private def maybeConvertToBlob(pathToTest: Try[Path]): Try[BlobPath] = {
+    pathToTest.collect { case blob: BlobPath => blob }
+  }
+
+  /**
+   * Under certain situations (and only on Terra), we want the VM running a TES task to have the ability to acquire a
+   * fresh sas token for itself. In order to be able to do this, we provide it with a precomputed endpoint it can use.
+   * The task VM will use the user assigned managed identity that it is running as in order to authenticate.
+   * We only return a value if at least one of the inputs is a BlobPath and //TODO flag specified in WDL.
+   * @param taskInputs The inputs to this particular TesTask. If any are blob files, the first  will be used to
+   *                   determine the storage container to retrieve the sas token for.
+   * @param pathGetter A function to convert string filepath into a cromwell Path object.
+   * @param blobConverter A function to convert a Path into a Blob path, if possible. Provided for testing purposes.
+   * @return A URL endpoint that, when called with proper authentication, will return a sas token.
+   *         Returns 'None' if one should not be used for this task.
+   */
+  def determineWSMSasEndpointFromInputs(taskInputs: List[Input],
+                                        pathGetter: String => Try[Path],
+                                        blobConverter: Try[Path] => Try[BlobPath] = maybeConvertToBlob): Option[String] = {
+    val shouldLocalizeSas = true //TODO: Make this a Workflow Option or come from the WDL
+    if (!shouldLocalizeSas) return None
+
+    // Collect all of the inputs that are valid blob paths
+    val blobFiles = taskInputs.collect{
+      case input if input.url.isDefined => BlobPathBuilder.validateBlobPath(input.url.get)
+    }.collect{
+      case valid: BlobPathBuilder.ValidBlobPath => valid
+    }
+    if(blobFiles.isEmpty) return None
```

Comment:
Rather than this early return, you can start the below `for` with
```
blobFile <- blobFiles.headOption
blob <- blobCOnverter(pathGetter(blobFile.toUrl))
```